[1단계 - 지하철 정보 관리 기능] 토리(천은정) 미션 제출합니다.

ezzanzzan opened the pull request:
안녕하세요 찰리 ! 👩🏻‍💻
지하철 미션을 함께하게 된 토리입니다 ˙ᵕ˙ 
이렇게 또 만나다니.. 정말정말정말 영광입니다 🙏🏻

레벨 2에 Spring을 처음 접하며 작동 되는 것에 만족하며 미션을 진행해오다
`객체지향적으로 설계하고 구현하는 것을 목표`로 잡고 미션을 진행하려고 하니 다시 처음으로 돌아간 것만큼 어려웠던 것 같아요. 

데이터베이스에서 표현하는 연관관계와 객체지향에서 표현하는 연관관계의 괴리감이 크게 다가왔습니다.
미션을 진행하며 하루종일 "객체지향적인게 뭔데..." 라는 생각을 떨칠 수 없었습니다 🥲
최대한 객체지향적으로 설계하고 구현하려고 노력했으나 현재 제 코드가 이런 괴리감에서 벗어난 코드인지.. 모르겠습니다 🥺

domain과 entity 사이의 간극을 어떻게 하면 좁혀서 비즈니스 로직을 구성할 수 있을지 찰리의 의견이 궁금합니다 🙇‍♀️

리뷰 잘 부탁드립니다 🍫 찰리 🍫

Cyma-s committed:
docs: 요구사항 정리

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: Line 생성 테스트-기능 구현

- 기존 코드 삭제

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: Station 생성 테스트-기능 구현

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
build: jgrapht 의존성 추가

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: Section 생성 테스트-기능 구현

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: Subway 생성 테스트-기능 구현

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
docs: API 명세서 추가

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
build: validation 의존성 추가

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
docs: API 요청/응답 예시 추가

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
docs: API 요구사항 수정

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: 노선 추가 API 테스트-기능 구현

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: 예외 처리 기능 구현

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: 노선 이름 중복 처리 테스트-기능 구현

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: 노선 조회 API 테스트-기능 구현

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
docs: 노선 목록 조회 API 예시 수정

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: 노선 목록 조회 API 테스트-기능 구현

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: 역 등록 API 테스트-기능 구현

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
refactor: 비즈니스 로직 책임 분리

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: Request Validation 기능 추가

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
feat: 역 삭제 API 테스트-기능 구현

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
test: StationTest 추가

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
test: SectionTest 추가

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
test: SectionsTest 추가

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
test: StationDao test 추가

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
test: SectionDaoTest 추가

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
test: LineDaoTest 추가

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
docs: API 예시 수정사항 반영

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Cyma-s committed:
refactor: 커스텀 예외 처리

Co-authored-by: ezzanzzan <eunzzann@gmail.com>

Gomding commented on .message.txt:
```diff
@@ -0,0 +1,3 @@
+
+
+Co-authored-by: ezzanzzan <eunzzann@gmail.com>
```
오 이런게 있었군요 😮 
오늘도 배워갑니다.. 🙇

ezzanzzan replied:
커밋내역에 빠질 수 없죠 🤩

Gomding commented on README.md:
```diff
@@ -1 +1,193 @@
+
+- [x] 이름을 갖는다.
+
+## API 명세서
```
API 명세서 까지.. 🤭 💯 💯 💯 💯 💯

ezzanzzan replied:
기술 관련 부분은 아니지만 개인적으로 궁금한 부분입니다 .. ! 

API 명세서을 현재 README 파일에 작성하고 있는데 README 파일이 길어져서 읽기 불편함이 있어서 어떤식으로 관리하는 것이 좋을지 계속 고민중입니다..🥲 

API 명세를 어떻게 관리할 수 있을지 찾아보다 postMan의 기능을 이용하는 방법도 고민해보았는데 따로 저장이 되는것이 아니라 [웹페이지로](https://documenter.getpostman.com/view/20982262/2s93m1b4se) 저장이 되어서요 🥺 

현재처럼 README에 관리하는 것이 좋을 지 아니면 포스트맨이나 다른 이외의 기능을 사용하는 것이 좋을 지 궁금합니다 ! 🧐

Gomding commented on src/main/java/subway/controller/LineController.java:
```diff
@@ -0,0 +1,40 @@
+    @PostMapping
+    public ResponseEntity<Void> createLine(@RequestBody LineRequest lineRequest) {
+        long id = lineService.save(lineRequest);
+        return ResponseEntity.created(URI.create("/lines/" + id)).build();
```
created 상태코드 반환과 Location 헤더 값 지정 👍

ezzanzzan replied:
데이터 추가에 대해 created 상태코드 반환과 Location 헤더 값을 지정해주었습니다. Location 헤더 값을 주는 이유가 해당 링크로 들어가 조회를 할 수 있기 위해서 라고 알고 있는데, 그렇다면 조회 가능한 링크에 대한 요청을 추가해주지 않았을 때는 따로 Location 헤더 값을 지정해주지 않는 것이 좋을까요?

또한 구간 추가와 같이 한 번에 두 가지의 값이 추가되는 경우 어떤 Location 값을 주어야 할 지 좋은 방법을 찾지 못해 결국 추가해주지 않았습니다.. 이 때 Location 헤더값이 없어도 괜찮을지 궁금합니다 ! 🧐

Gomding replied:
좋은 고민이네요 👍 

> Location 헤더 값을 주는 이유가 해당 링크로 들어가 조회를 할 수 있기 위해서 라고 알고 있는데, 그렇다면 조회 가능한 링크에 대한 요청을 추가해주지 않았을 때는 따로 Location 헤더 값을 지정해주지 않는 것이 좋을까요?

조회 가능한 API 가 없다면 따로 Location 헤더 값을 지정해주지 않아도 된다고 생각해요~
201 Created 상태 코드의 응답에서 Location 헤더 값은 `새로 만들어진 리소스는 여기에 있어` 를 의미하기 때문에
만약 조회할 수 있는 API 가 없다면 Location 값을 지정하지 않는것이 혼란을 없앨 수 있겠죠 :)

> 구간 추가와 같이 한 번에 두 가지의 값이 추가되는 경우 어떤 Location 값을 주어야 할 지 좋은 방법을 찾지 못해 결국 추가해주지 않았습니다.. 이 때 Location 헤더값이 없어도 괜찮을지 궁금합니다 ! 🧐

구간 추가와 같은 경우는 단일 구간 조회에 대한 API 가 없기도 하고 추가할 때 구간이 update + insert 될 수 있으므로 애매하여 나온 고민으로 보이네요.
이 경우는 클라이언트와 협의를 해볼 수 있겠네요.
- 노선 조회 API를 Location 값에 넣어 노선에 포함된 구간들을 조회할 수 있도록 응답
- 애매하니 따로 Location 헤더 값을 지정하지 않고 응답 😄 

토리는 어떻게 생각하시나요?

ezzanzzan replied:
구간을 따로 조회할 수 있는 API가 존재하지 않고 있어서 헤더 값을 지정하지 않고 있었는데, 구간 조회를 할 수 있는 API는 없지만 노선 조회 API를 통해 **변경된 점을 확인할 수 있도록 노선 조회 API를 Location 값에 넣어주는 것**이 좋을 것 같다고 생각이 들었습니다 !

구간 추가 후 해당 구간이 존재하는 노선을 조회하도록 Location 헤더 값을 지정해주는 방향으로 리팩토링 하였습니다 🙇‍♀️

Gomding commented on src/main/java/subway/service/LineService.java:
```diff
@@ -0,0 +1,81 @@
+        } catch (EmptyResultDataAccessException e) {
+            return false;
+        }
+    }
```
특정 이름을 가진 라인이 존재하는지 확인하고 싶다면 EXISTS 쿼리를 사용해볼 수 있어요 :)

```sql
SELECT EXISTS(SELECT * from LINE WHERE name = ?)
```

현재는 `SELECT * FROM line WHERE name = ?` 를 실행하고 있는데
해당 쿼리는 1건을 찾더라도 계속 해당 이름을 가진 녀석들이 더 있는지 찾아다닙니다 👀 👀 
즉 전체를 스캔해보겠죠? (index 가 걸려있다면 조금 다르겠지만요!)

만약 Exists 를 사용할 수 없는 경우라면 `Limit 1`을 줘서 `1개 발견하면 탐색을 멈춰`라고 할 수 있습니다~

```sql
SELECT * FROM line WHERE name = ? LIMIT 1
```

ezzanzzan replied:
오 이런 방법이 있었군요...! 특정 이름이 존재하는 컬럼이 존재하는지 확인하는 부분에 대한 쿼리를 확인하여 EXISTS 로 변경해주었습니다 !! 🙇‍♀️

Gomding commented on src/main/java/subway/exception/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,31 @@
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new ErrorResponse(errorMessage));
+    }
+
+    @ExceptionHandler({DuplicateException.class, InvalidException.class})
+    public ResponseEntity<ErrorResponse> handleDuplicateException(CustomException e) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(new ErrorResponse(e.getMessage()));
+    }
+
+    @ExceptionHandler(NotFoundException.class)
+    public ResponseEntity<ErrorResponse> handleNotFoundException(NotFoundException e) {
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(new ErrorResponse(e.getMessage()));
+    }
```
상태 코드가 더 생긴다면 핸들링 메서드 수가 계속 늘어나겠어요! 😭 

아래와 같이 CustomException 이 HttpStatus 를 필드로 가진다면 CustumException 만 핸들링할 수 있지 않을까요!

```java
public class CustumException extends RuntimeException {
    private final String message;
    private final HttpStatus httpStatus;

    public CustomException(final String message, final HttpStatus httpStatus) {
        this.message = message;
        this.httpStatus = httpStatus;
    }
}
```

ezzanzzan replied:
그렇지 않아도 CustomException으로 처리하는 메소드가 늘어날 때마다 새로운 핸들링 메서드를 추가해주어야 해서 좋은 방법이 없을까 고민하던 중이었는데 미션 구현 시 시간이 부족해서 넘어갔던 부분이었습니다 🥲

추천해주신 방법을 통해서 HttpStatus를 필드로 가지게 한 뒤 CustomException을 상속 받는 클래스들에서 HttpStatus를 지정해주니 코드가 훨씬 깔끔해지고 중복되는 코드들도 줄일 수 있었습니다 👍

Gomding commented on src/main/java/subway/exception/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,31 @@
+
+@RestControllerAdvice
+public class GlobalControllerAdvice {
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<ErrorResponse> handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
```
MethodArgumentNotValidException 핸들링 👍

Gomding commented on src/main/java/subway/exception/CustomException.java:
```diff
@@ -0,0 +1,14 @@
+package subway.exception;
+
+public class CustomException extends RuntimeException {
```
취향 차이인 부분이지만

CustomException 이라는 이름은 정보가 반만 들어있겠군요 :)
(기존에 존재하지 않던 커스텀 예외다~ 라는 정보만 알 수 있었습니다 😄 )

쓰임을 보면 Client 에게 전달될 예외로 보여요 따라서 `ClientException` 
또는 지하철 애플리케이션의 예외이니 `SubwayException` 등으로 지어보면 어떨까요!

ezzanzzan replied:
CustomException 이라는 이름으로 커스텀 예외라는 것을 명시해주고 추후 Exception을 상속받는 부분에서 네이밍을 해주면 될 것이라고 생각했었는데, 

찰리의 코멘트를 보고 `SubwayException`과 같이 쓰임에 따라 이름을 명시해주는 것이 훨씬 알아보기 좋다고 생각해 해당 방향으로 리팩토링 하였습니다 ˙ᵕ˙ !!!

Gomding commented on src/main/resources/schema.sql:
```diff
@@ -1,14 +1,32 @@
-    primary key(id)
+    id      BIGINT       NOT NULL AUTO_INCREMENT,
+    name    VARCHAR(255) NOT NULL UNIQUE,
+    line_id BIGINT       NOT NULL,
```
Station 이 `line_id` 를 알고있도록 설계하신 의도가 궁금해요! 😄 

`name` 컬럼이 unique 라서 같은 이름 을 가진 역이 생길 수 없을 것 같아요
만약 환승역이 생긴다면 line_id 만 다른 역이 생성될 텐데
토리는 어떻게 생각하시나요 🤔

ezzanzzan replied:
기존 생각은 Station이 여러 개의 라인을 가질 수 있다고 생각하고 설계했습니다. 
그래서 이렇게 테이블 설계를 했었는데 코멘트를 보고 다시 확인해보니 말이 되지 않는 구조였네요.. 시간에 쫓겨 제대로 확인하지 못했던 부분이 많은 것 같습니다 🥲

Station 테이블에서 line_id를 제거하고 name은 UNIQUE 그대로 두어 Station 테이블은 아이디와 이름만 가지고 있도록 리팩토링 하였습니다 !!

Gomding commented on src/main/java/subway/service/StationService.java:
```diff
@@ -0,0 +1,130 @@
+        List<Section> newSections = line.getSectionsByList();
+        originalSections.removeAll(newSections);
+        Section deletedSection = originalSections.get(0);
+        StationEntity deletedUpStation = stationDao.findByName(deletedSection.getUpStation().getName());
+        StationEntity deletedDownStation = stationDao.findByName(deletedSection.getDownStation().getName());
+
+        sectionDao.delete(new SectionEntity(
+                deletedUpStation.getId(),
+                deletedDownStation.getId(),
```
해당 로직을 봤을 때 station의 id를 얻기 위해 DB에서 조회해오는 것으로 보여요 😃 

Station 객체가 id를 가지는것에 대해서는 어떻게 생각하시나요?

ezzanzzan replied:
해당 부분에 대해 슬랙으로 얘기를 나눴던 부분과 크루들과도 더 얘기를 나눠보면서 Station 객체가 id를 가지고 있는 방향으로 리팩토링 하고, station의 id를 조회하는 쿼리 등 불필요한 쿼리들을 최대한 줄여보았습니다 🙇‍♀️

Gomding commented on src/main/java/subway/service/StationService.java:
```diff
@@ -0,0 +1,130 @@
+    private void saveNewSections(final long lineId, final List<Section> originalSections, final Line line) {
+        List<Section> newSections = line.getSectionsByList();
+        newSections.removeAll(originalSections);
+        for (Section newSection : newSections) {
+            StationEntity newUpStation = stationDao.findByName(newSection.getUpStation().getName());
+            StationEntity newDownStation = stationDao.findByName(newSection.getDownStation().getName());
+
+            sectionDao.save(new SectionEntity(
+                    newUpStation.getId(),
+                    newDownStation.getId(),
+                    lineId,
+                    newSection.getDistance())
+            );
+        }
```
batchInsert 를 사용하면 쿼리를 여러번 실행하지 않아도 되겠어요 :)

ezzanzzan replied:
batchInsert를 사용했을 때와 여러번 insert 해주었을 때의 쿼리 속도(성능?) 차이를 찾아보며 (현재는 최대 2개의 리스트만 존재하긴 하지만.. ) batchUpdate 를 사용하는 방식으로 리팩토링 하였습니다 !

다만 batchUpdate를 처음 사용해 보는터라 제대로 사용된 것인지 궁금합니다.. 🥺
```java
    public void batchSave(List<SectionEntity> sectionEntities) {
        final String sql = "INSERT INTO section "
                + "(`up_station_id`, `down_station_id`, `line_id`, `distance`)"
                + " VALUES (?, ?, ?, ?)";

        jdbcTemplate.batchUpdate(sql, new BatchPreparedStatementSetter() {
            @Override
            public void setValues(PreparedStatement ps, int i) throws SQLException {
                ps.setLong(1, sectionEntities.get(i).getUpStationId());
                ps.setLong(2, sectionEntities.get(i).getDownStationId());
                ps.setLong(3, sectionEntities.get(i).getLineId());
                ps.setInt(4, sectionEntities.get(i).getDistance());
            }

            @Override
            public int getBatchSize() {
                return sectionEntities.size();
            }
        });
    }
```

Gomding replied:
잘 사용해주셨어요 👍

Gomding commented on src/main/java/subway/service/StationService.java:
```diff
@@ -0,0 +1,130 @@
+        this.lineDao = lineDao;
+    }
+
+    public void save(final StationCreateRequest stationCreateRequest) {
```
StationService 의 save 메서드는 역을 생성하는 것보다 구간(Section)을 생성하는것이 중심으로 보이는데 맞을까요!

`schema.sql` 에서도 비슷한 느낌을 느낌을 받았지만

역(Station)을 생성하는 것과 구간(Section)을 생성하는 것이 나눠져있으면 어떨까 생각합니다 :)

- 역이 먼저 등록 -> 노선에 역을 구간으로 등록
- 지하철 노선이라는 도메인에서 역(station) 은 노선에 종속적인 존재가 아니라고 생각해요 :)
- 환승이 가능하기 때문이죠!

ezzanzzan replied:
schema.sql에서 Station 테이블을 수정하며 전체적인 API에 대해 다시 생각해보고 리팩토링하였습니다.

기존의 역 추가, 삭제로만 구성되어 있던 API 를 아래와 같이
### 역 API
| Method | URI       | Description |
|--------|-----------|-------------|
| POST   | /stations | 역 추가        |

### 구간 API
| Method | URI       | Description |
|--------|-----------|-------------|
| POST   | /sections | 노선에 역 추가    |
| DELETE | /sections | 노선에 역 삭제    |

역을 추가할 때는 이름을 통해 역만 추가하고, 
구간 이라는 API를 추가하여 노선에 역을 추가하고 노선에 역을 삭제하도록 변경하였습니다...!!

Gomding commented on src/test/java/subway/integration/LineIntegrationTest.java:
```diff
@@ -56,59 +55,20 @@ void createLineWithDuplicateName() {
-                .map(LineResponse::getId)
-                .collect(Collectors.toList());
-        assertThat(resultLineIds).containsAll(expectedLineIds);
+        assertThat(response.statusCode()).isEqualTo(HttpStatus.BAD_REQUEST.value());
```
예외 메시지도 검증해볼 수 있겠군요 :)

ezzanzzan replied:
이 부분도 시간관계상 넘어갔던 부분이었습니다 🥲

전체적으로 테스트 코드를 점검하며 예외 처리 테스트 코드 및 예외 메시지 검증 코드를 추가하였습니다 !! 🙇‍♀️

Gomding commented on src/test/java/subway/integration/LineIntegrationTest.java:
```diff
@@ -138,53 +98,42 @@ void getLine() {
         assertThat(resultResponse.getId()).isEqualTo(lineId);
     }
```
조회한 lineId 만 검증해도 될지 의문이 들었어요! 🤔 
응답받은 다른 값들도 검증해볼 수 있겠군요

ezzanzzan replied:
해당부분과 전체적으로 부족한 테스트 부분에 대해서 추가해주었습니다 !!🙇‍♀️

Gomding commented on src/test/java/subway/integration/LineIntegrationTest.java:
```diff
@@ -138,53 +98,42 @@ void getLine() {
+        List<Long> expectedLineIds = Stream.of(createResponse1, createResponse2)
+                .map(it -> Long.parseLong(it.header("Location").split("/")[2]))
+                .collect(Collectors.toList());
+        List<Long> resultLineIds = response.jsonPath().getList(".", LineResponse.class).stream()
+                .map(LineResponse::getId)
+                .collect(Collectors.toList());
```
응답에서 Response 배열을 잘 추출해주셨네요 💯

Gomding reviewed (changes requested):
안녕하세요 토리!
레벨2에서 다시 만나 반갑습니다~ 🙌 
복잡한 요구사항임에도 전체적으로 깔끔하게 구현해주셨네요 💯 
설계해주신 부분에 궁금한 부분이 있어 몇가지 코멘트 남겼습니다!

궁금한 점 있으시면 언제든 DM 남겨주세요~!!

Gomding commented:
질문을 깜빡했네요 😭 

저는 도메인 설계를 엄청 잘해주셨다고 생각했어요 💯 💯 💯 💯 
굉장히 객체지향적으로 잘 만들었습니다 ㅎㅎ

> domain과 entity 사이의 간극을 어떻게 하면 좁혀서 비즈니스 로직을 구성할 수 있을지 찰리의 의견이 궁금합니다 🙇‍♀️

질문의 의도를 이해하지 못했어요 ㅠㅠ
domain과 entity 사이의 간극에는 어떤 부분이 있었을까요? 😄

ezzanzzan committed:
docs: 리팩토링 목록 추가

ezzanzzan committed:
style: 불필요한 공백 제거

ezzanzzan committed:
refactor: ExceptionHandler 처리 방식 변경

ezzanzzan committed:
style: 코드 컨벤션 적용

ezzanzzan committed:
docs: 기능 요구사항 수정

ezzanzzan committed:
refactor: 역 추가 기능 리팩토링

- 역 추가 기능과 구간 추가 기능 분할
- 역 추가 API 추가
- 노선 추가 API 수정
- station table 수정

ezzanzzan committed:
refactor: 중복된 데이터 체크하는 로직 변경

- DB 내 데이터 EXISTS query료 확인하도록 수정
- 역 추가 시 중복 확인 로직 추가

ezzanzzan committed:
feat: 예외 처리 클래스 추가

- ErrorCode 수정
- NoSuchException 예외 처리 클래스 추가

ezzanzzan committed:
refactor: Station 클래스 리팩토링

- Station 객체가 id를 갖도록 리팩토링
- 역 추가 API, 역 삭제 API 변경
- StationService, SectionService 분리

ezzanzzan committed:
refactor: Line 리팩토링

- deleteStation -> deleteSection 네이밍 변경
- 예외 처리 메세지 변경

ezzanzzan committed:
refactor: StationService saveNewSections 메소드 내부 로직 변경

- Section Dao batchSave 메소드 기능 구현
- batchUpdate 중복 쿼리 제거

ezzanzzan committed:
test: LineDaoTest 추가

ezzanzzan committed:
test: SectionDaoTest 추가

ezzanzzan committed:
test: StationDaoTest 추가

ezzanzzan committed:
test: 도메인 테스트 추가

ezzanzzan committed:
test: 통합 테스트 수정

ezzanzzan committed:
test: 테스트 시 예외 메시지 추가

ezzanzzan committed:
style: 컨벤션 적용

ezzanzzan committed:
test: 테스트 리팩토링

- 테스트 추가
- assertAll로 변경

Gomding commented on src/test/java/subway/integration/SectionIntegrationTest.java:
```diff
@@ -0,0 +1,147 @@
+                .when()
+                .post("/stations")
+                .then().log().all()
+                .extract();
```
테스트를 위한 사전 데이터 준비를 위해 중복되는 코드가 보이네요 :)
테스트 가독성을 위해 다음과 같이 사용해보는건 어떻게 생각하세요?

```java
public class StationStep {
    private StationStep() { }

    public static Long 역을_생성한다(StationCreateRequest createRequest) {
        ExtractableResponse<Response> response = RestAssured.given().log().all()
                .body(createRequest)
                .contentType(MediaType.APPLICATION_JSON_VALUE)
                .when()
                .post("/stations")
                .then().log().all()
                .extract();

        String location = response.header("Location");
        Long stationId = Long.parseLong(location.substring(location.lastIndexOf("/") + 1));
    }
}

public class SectionIntegrationTest extends IntegrationTest {

    @BeforeEach
    public void setUp() {
        Long 잠실역_ID = StationStep.역을_생성한다(createRequest1);
        Long 잠실새내역_ID = StationStep.역을_생성한다(createRequest2);
    }
```

ezzanzzan replied:
계속 준비해줘야할 데이터가 늘어나 어떤식으로 관리해야 할 지 고민이었는데 코멘트에 남겨주신 방법을 통해 테스트 코드 내의 중복되는 코드에 대해서 최대한 줄여보았습니다 ˙ᵕ˙ 훨씬 가독성이 좋아진 걸 느꼈어요 👍

Gomding commented on src/main/java/subway/service/SectionService.java:
```diff
@@ -0,0 +1,134 @@
+            return stationDao.findByName(stationName);
+        } catch (Exception e) {
+            System.out.println(e.getMessage());
+            throw new NoSuchException(ErrorCode.NO_SUCH_STATION);
```
NoSuchException 은 BadRequest(400) 상태 코드로 응답되는 것으로 보여요

리소스를 찾지 못했을 때 NotFound(404) 상태코드로 응답하는것은 어떻게 생각하시나요?

ezzanzzan replied:
`NoSuchException`은 해당하는 데이터를 데이터베이스에서 찾지 못했을 때 발생하는 Exception으로 클라이언트에서 요청한 부분에 오류가 있다고 생각해서 BadRequest(400) 상태 코드로 응답하였습니다 ! 

`NotFound(404)` 상태 코드는 **API url에 관련된 예외 상황**이 발생했을 때 사용되는 상태 코드라고 생각했습니다. 🥺

저는 **클라이언트에서 잘못된 파라미터(정보)를 가지고 요청을 보냈을 때** 이 데이터를 찾지 못해서 발생하는 오류에는 `BadRequest(400)` 상태 코드로 응답되는 것이 좀 더 상황을 이해하기 쉽다고 생각했습니다 ! 

NoSuchException이라는 class 네이밍 때문에 발생하게 된 모호함일까요 .. 🥲

찰리는 400 상태코드와 404 상태 코드를 어떤 기준으로 사용하고 계시나요 !?!? 🧐

Gomding replied:
기준을 확실하게 잡고 계시네요 👍 
저는 NOT_FOUND 로 상태 코드를 표현하는 편입니다 ㅎㅎ 하지만 절대 정답이 아니고 
토리의 말씀대로 `요청한 url 자체가 없을 때 사용되는 상태 코드`로 정할 수 있어요~

이 부분은 클라이언트와 협의에 따라 다를 수 있는 부분이라 생각해요!

Gomding commented on src/main/java/subway/service/SectionService.java:
```diff
@@ -0,0 +1,134 @@
+import subway.exception.NoSuchException;
+
+@Service
+@Transactional
```
`@Transactional` 애너테이션을 Service 전역에 적용되도록 하면 의도하지 않게 트랜잭션 범위가 잡힐 수 있어요 😢 
이는 긴 범위의 트랜잭션을 유발할 수 있습니다 :)
필요한 부분에서만 트랜잭션을 선언해보는 것은 어떨까요?

ezzanzzan replied:
`@Transactional` 어노테이션을 페어 미션하며 처음 사용하며 
> 필요한 부분에만 `@Transactional` 어노테이션을 적용시킬 때 추후 리팩토링 과정이나 기능을 추가할 때 `@Transactional` 어노테이션이 추가되어야 할 부분에 실수로 추가하지 않아 문제가 발생할 수도 있다. -> 서비스 전역에 `@Transactional` 어노테이션을 추가하고 조회와 같이 필요하지 않은 부분에는 제거하자

라는 방향으로 페어와 진행을 했었습니다 !

저도 꼭 필요한 부분에 트랜잭션이 걸리지 않는 것보다 위와 같은 방법이 안전하다고 생각했습니다 🥺

다만 현재 코드를 확인해보니 SectoinService 내에서는 불필요한 부분에도 트랜잭션이 사용되고 있었네요 제대로 확인하지 못하고 놓쳤던 부분입니다 🥲
트랜잭션이 불필요한 부분에는 읽기 전용 모드로 설정해주는 방향으로 리팩토링을 진행했습니다 !

찰리는 `@Transactional` 어노테이션을 사용하실 때 두 가지 방법 중 어떤 방식으로 사용하시는 편인가요 !? 🧐
1. 전체적으로 `@Transactional` 어노테이션을 적용시킨 뒤 불필요한 부분(메소드)에서 읽기 전용 모드로 선언
2. 필요한 부분에서만 `@Transactional` 어노테이션 선언

ezzanzzan replied:
`@Transactional` 어노테이션을 페어 미션하며 처음 사용하며 
> 필요한 부분에만 `@Transactional` 어노테이션을 적용시킬 때 추후 리팩토링 과정이나 기능을 추가할 때 `@Transactional` 어노테이션이 추가되어야 할 부분에 실수로 추가하지 않아 문제가 발생할 수도 있다. -> 서비스 전역에 `@Transactional` 어노테이션을 추가하고 조회와 같이 필요하지 않은 부분에는 제거하자

라는 방향으로 페어와 진행을 했었습니다 !

저도 꼭 필요한 부분에 트랜잭션이 걸리지 않는 것보다 위와 같은 방법이 안전하다고 생각했습니다 🥺

트랜잭션이 불필요한 부분에는 `@Transactional(readOnly = true)`으로 설정해주는 방향으로 리팩토링을 하려고 했습니다.
현재 코드에서는 조회만 하는 기능은 LineService 에서만 사용하고 있어 해당 부분을 제외하고는 트랜잭션이 붙는 것이 맞다고 생각이 들었습니다 .. ! 🥲

---

혹시 제가 트랜잭션에 대해 잘못된 이해를 하고 있는걸까요 ?

찰리는 `@Transactional` 어노테이션을 사용하실 때 두 가지 방법 중 어떤 방식으로 사용하시는 편인가요 !? 🧐
1. 전체적으로 `@Transactional` 어노테이션을 적용시킨 뒤 불필요한 부분(메소드)에서 읽기 전용 모드로 선언
2. 필요한 부분에서만 `@Transactional` 어노테이션 선언

Gomding replied:
오히려 잘 이해하고 사용하고 계십니다 🤗

다만 가끔 특수한 케이스로 트랜잭션이 길어지는 경우가 발생할 수 있어요
예를들면 Service 에서 사용하는 `외부 api` 의 응답이 지연되는 경우 트랜잭션이 걸려있으면 그만큼 지연될 수 있겠죠? 😉 
보통 트랜잭션을 최대한 짧게 잡는편이 좋다고 합니다 :) 왜 그럴까요? 🤔

Gomding commented on src/test/java/subway/domain/SectionsTest.java:
```diff
@@ -0,0 +1,157 @@
+        assertThat(stations).contains(station1, station2, station3);
+    }
+
+    @DisplayName("새로운 구간을 추가한다.")
+    @Test
+    void addSection() {
+        // given
+        Station upStation = new Station(4L, "종합운동장역");
+        Section newSection = new Section(upStation, station3, 2);
+
+        // when
+        sections.addSection(newSection);
```
해당 테스트에서 검증하는 것은 `역 사이에 새로운 역을 추가할 수 있다~` 로 보여요 😄 

실제로 리뷰하면서 아래의 요구사항에 대한 테스트를 찾아보고 있었어요 ㅎㅎ
`새로운 구간을 추가한다.` 라는 이름으로는 해당 테스트에서 검증하는 부분을 충분히 전달하고 있지 않다고 생각돼요 😄 
토리는 어떻게 생각하시나요?

```
- 노선에 등록되는 역의 위치는 자유롭게 지정할 수 있어야 합니다.
  - A-B-C 역이 등록되어 있는 노선에 D역을 등록할 경우 C역 다음에만 등록할 수 있는게 아니라
  - A역과 B역 사이에도 등록이 가능하고, A역 앞에도 등록을 할 수 있어야 합니다.
```

ezzanzzan replied:
SectionsTest 코드 내부에서 BeforeEach 어노테이션으로 테스트를 시작하기 전
`잠실 - 잠실새내역`, `잠실새내역 - 봉천역` 구간을 추가 된 상태로 테스트를 시작하도록 구현하였습니다 !

```java
@DisplayName("새로운 구간을 추가한다.")
@Test
void addSection() {
    // given
    Station upStation = new Station(4L, "종합운동장역");
    Section newSection = new Section(upStation, station3, 2);

    // when
    sections.addSection(newSection);

    // then
    assertAll(
            () -> assertThat(sections.getStations()).hasSize(4),
            () -> assertThat(sections.getSections().get(1)).isEqualTo(newSection)
    );
}

...

@DisplayName("새로운 역이 상행 종점으로 추가된다.")
@Test
void addSectionFirstStation() {
    // given
    Station newUpStation = new Station(4L, "베로역");
    Section newSection = new Section(newUpStation, station1, 10);

    // when
    sections.addSection(newSection);

    // then
    assertThat(sections.getSections().get(0)).isEqualTo(newSection);
}

@DisplayName("새로운 역이 하행 종점으로 추가된다.")
@Test
void addSectionLastStation() {
    // given
    Station newDownStation = new Station(4L, "베로역");
    Section newSection = new Section(station3, newDownStation, 10);

    // when
    sections.addSection(newSection);

    // then
    assertThat(sections.getSections().get(2)).isEqualTo(newSection);
}
```

1. `@DisplayName("새로운 구간을 추가한다.")`
메서드 내부에서는 `잠실새내역 - 봉천역` 사이에 새로운 역을 추가했을 때의 경우를

2. `@DisplayName("새로운 역이 상행 종점으로 추가된다.")`
메서드 내부에서는 잠실역 앞에 새로운 역을 추가했을 때의 경우를

3. `@DisplayName("새로운 역이 하행 종점으로 추가된다.")`
메서드 내부에서는 봉천역 다음으로 새로운 역을 추가했을 때로...

답변을 작성하며 제 테스트 코드 내의 DisplayName이 테스트의 내용을 제대로 전달하지 못하고 있다는걸 느꼈어요 🥺 DisplayName이 명확하지 않아 요구사항에 대한 테스트를 찾는데 어려움이 있으셨을 것 같아요........🥲

DisplayName을 전체적으로 확인하고 좀 더 명확하게 테스트의 내용을 전달할 수 있도록 리팩토링 하였습니다 🙇‍♀️

Gomding replied:
추가로 고민해주신 부분까지 훌륭하네요 💯 

개인적인 취향으로는 테스트를 읽다고면 현재 역이 어떻게 연결됐는지 확인해야하고 
검증할 때 구간이 어떻게 변경돼야하는지 주석으로 설명해주는 것도 좋다고 생각해요 ㅎㅎ

```
@DisplayName("새로운 B-D 구간이 B-C 사이에 추가된다.")

    // given
    // A - B - C

    // then
    // A - B - D - C
```

Gomding commented on src/main/java/subway/service/SectionService.java:
```diff
@@ -0,0 +1,134 @@
+        }
+    }
+
+    private Line makeLine(Long lineId) {
+        LineEntity lineEntity = lineDao.findById(lineId);
+        List<Section> sections = makeSections(lineId);
+
+        return Line.of(lineEntity.getName(), sections);
+    }
```
Sections를 담고있는 Line 을 조회할 수 있도록 해주셨네요 👍 

실제 행위는 `make` 보다는 `find`라는 조회에 가깝지 않을까요? 🤔

ezzanzzan replied:
lineId를 통해 Dao 에서 해당되는 값들을 조회한 뒤 Line 객체로 만드는 행위를 한다고 생각해서 `make` 라는 네이밍을 하게 되었습니다.
`find` 라는 네이밍은 말 그대로 기존에 있던 데이터를 찾는 느낌이 강하다는 생각이 들었습니다 😔

`make` 라는 이름이 많이 어색할까요 ?! 
네이밍에 대한 고민을 하다 `converTo...ByEntity` 라는 네이밍으로 변경해보았는데 혹시 해당 네이밍에 대해서는 어떻게 생각하시나요 !! 🧐

Gomding replied:
앗 서비스 입장에서는 단순히 Line을 조회한다~ 정도만 알고 있으면 된다고 생각해요 :)
makeLine 메서드를 호출하는 쪽에서도 `LineDao 에서 조회된 entity 를 Line 객체로 변환한다` 라는 정보는 크게 신경쓰지 않는 정보라고 생각합니다 🤔

Gomding commented on src/main/java/subway/service/SectionService.java:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    private List<Section> makeSections(Long id) {
+        List<SectionEntity> sectionEntities = sectionDao.findByLineId(id);
+        Map<Long, String> stationEntities = stationDao.findAll().stream()
```
`Section 을 조회할 때 join 으로 station 정보를 함께 조회하기` 
**VS**
`Section 정보를 조회해오고 Station 정보를 모두 가져와서 조립하기` 

토리는 어떤 방식을 선호하시나요?

ezzanzzan replied:
저는 후자의 방식을 선호하는 편입니다 !
join을 사용하게 되면 해당 결과값을 받는 객체도 새로 생성되어야 하고 join 쿼리를 사용하는게 조금 더 복잡하게 느껴지기도 해서요 🥲

찰리는 어떤 방식을 선호하시는 편인가요 !?

Gomding replied:
저는 DB 조회를 줄이기위해 join 쿼리를 사용하는것을 선호합니다 ㅎㅎ
(성능을 위해서요!)
다만 토리의 말씀대로 join 쿼리는 유지보수하기에 힘이듭니다
복잡하게 작성된 join 쿼리를 3개월 뒤에 내가 유지보수 하기 쉬울까? 라는 의문을 가져보는 자세도 좋습니다 💯 💯 💯 

이부분도 트레이드 오프라고 생각해요 ㅎㅎ

Gomding commented on src/main/java/subway/service/SectionService.java:
```diff
@@ -0,0 +1,134 @@
+                )).collect(Collectors.toList());
+    }
+
+    private void saveNewSections(long lineId, List<Section> originalSections, Line line) {
+        List<Section> newSections = line.getSectionsByList();
+        newSections.removeAll(originalSections);
+
+        List<SectionEntity> newSectionEntities = newSections.stream()
+                .map(section ->
+                        new SectionEntity(
+                                section.getUpStation().getId(),
+                                section.getDownStation().getId(),
+                                lineId,
+                                section.getDistance()))
+                .collect(Collectors.toList());
+
+        sectionDao.batchSave(newSectionEntities);
+    }
+
+    private void deleteOriginalSection(long lineId, List<Section> originalSections, Line line) {
+        if (originalSections.isEmpty()) {
+            return;
+        }
+        List<Section> newSections = line.getSectionsByList();
+        originalSections.removeAll(newSections);
+        Section deletedSection = originalSections.get(0);
+        sectionDao.delete(new SectionEntity(
+                deletedSection.getUpStation().getId(),
+                deletedSection.getDownStation().getId(),
+                lineId,
+                deletedSection.getDistance())
+        );
+    }
```
새로운 구간을 등록하고 이전에 이어져있던 구간은 삭제해주는 방식을 사용하고 계시는군요 :) 

`잠실역 - 잠실새내역` -> 사이에 모시깽이역 추가 -> `잠실역 - 모시깽이역, 모시깽이역 - 잠실새내역` 

아래와 같이 진행해보면 같은 결과가 나올 것 같은데 토리는 어떻게 생각하시나요? 😄 
- 노선에 있는 구간을 전체 삭제
- 업데이트된 노선의 구간을 전체 저장

```java
sectionDao.deleteByLineId(lineId);
sectionsDao.batchSave(line.getSectionsByList());
```

ezzanzzan replied:
사실 처음에는 기존에 바뀐 부분을 찾을 필요 없이 모두 삭제하고 저장하는 방식으로 사용할까 라는 생각으로 라인에 해당하는 모든 구간을 삭제한 뒤 업데이트를 진행하려고 했습니다 !
로직 상에서 변경된 부분만을 구할 방법을 생각했고, 모든 구간에 대한 삭제, 추가가 일어나는 것보다는 변경이 있는 리소스에 대해서만 삭제, 추가를 해주는 것이 성능 면으로 더 좋다는 생각에 현재와 같은 방식으로 구현했습니다. 🥺

다만 지하철 노선도는 데이터가 많은 경우도 아니고, 추가,삭제가 자주 일어나는 기능이 아니라는 생각도 들던 와중에 찰리의 코멘트를 보고 삭제할 리소스, 추가할 리소스를 찾는 긴 코드를 두 줄로 줄일 수 있다는 걸 눈으로 보고 나니 코드를 깔끔하게 관리하고 싶어서 찰리가 추가해주신 방법을 더 쓰고 싶어져 추천해주신 방식으로 리팩토링 해보았습니다 👍

Gomding commented on src/main/java/subway/domain/Section.java:
```diff
@@ -0,0 +1,78 @@
+        }
+    }
+
+    public void validateDistance(int distance) {
+        if (this.distance <= distance) {
+            throw new InvalidException(ErrorCode.INVALID_DISTANCE);
+        }
+    }
```
1. 해당 메서드의 접근자는 public 인 이유가 궁금해요!
2. 구간을 추가하는데 거리에 대한 제약사항을 Sections가 아닌 Section 이 알고있는 구조가 어색하게 느껴져요 🤔

ezzanzzan replied:
1. Section 내부에서만 사용되는 메서드인데 접근제어자가 public 이었네요 🥲 리팩토링 과정에서 다시 private로 수정하였습니다 !

2. 거리에 대한 검증 로직에 모두 Section 내부의 값이 사용되어 getter로 값을 꺼내서 사용하는 것보다 내부 필드로 가지고 있는 Section 내에서 검증하는 것이 어색하다는 생각을 하지 못했었습니다 🥺 코멘트를 보고 생각해보니 추가할 때 기존 구간과 새로운 구간의 거리를 검증하는 로직은 Sections에 있는 것이 맞다는 생각이 들었습니다. 또 어떤 구간(기존 구간) 사이에 들어가는지 판단이 되고 난 뒤에 바로 검증하는 진행 순서가 맞다고 생각해 현재 Sections 내부 코드를 아래와 같이 변경하였습니다 !
```java
private void removeExistSectionAddNewSection(Section newSection, Section existSection) {
    validateDistance(newSection.getDistance(), existSection.getDistance());

    int currentSectionIndex = sections.indexOf(existSection);
    Section dividedSection = existSection.getDividedSection(newSection);

    sections.remove(currentSectionIndex);
    sections.add(currentSectionIndex, newSection);
    sections.add(currentSectionIndex + 1, dividedSection);
}

private void validateDistance(int newDistance, int existDistance) {
    if (existDistance <= newDistance) {
        throw new InvalidException(ErrorCode.INVALID_DISTANCE);
    }
}
```

Gomding reviewed (changes requested):
안녕하세요 토리!
피드백 잘 반영해주셨네요 👍 
질문 주신 부분에 대한 답변과 함께 몇가지 코멘트 남겼습니다 :)

추가로 궁금한 점 있으면 언제든 편하게 DM 남겨주세요~

ezzanzzan committed:
docs: 리팩토링 목록 추가

ezzanzzan committed:
refactor: 구간 추가 API 요청 시 응답되는 헤더 값 변경

- 구간 추가 API Location 헤더 값에 해당 노선 조회할 수 있는 url 지정

ezzanzzan committed:
refactor: SectionService 내부 로직 변경

- 구간 추가, 삭제 시 DB에 저장, 삭제하는 로직 전체 데이터 삭제 후 batchInsert 하는 방식으로 변경

ezzanzzan committed:
fix: 구간 추가 오류 수정

- 두 역 사이에 하행 방향으로 추가하는 경우 삭제되지 기존 구간 삭제되지 않던 오류 수정

ezzanzzan committed:
refactor: validateDistance 메서드 리팩토링

- 접근제어자 변경
- Sections 클래스로 이동

ezzanzzan committed:
refactor: SectionService 내 객체 생성 메소드 네이밍 변경

- make -> convertToLineByEntity

ezzanzzan committed:
refactor: Dao 내 불필요한 코드 제거 및 테스트 코드 리팩토링

ezzanzzan committed:
test: 테스트 코드 리팩토링

- 테스트 코드 내부 DisplayName 명확하게 수정
- `@Nested` 어노테이션을 통해 계층 구조로 테스트 코드 가독성 증진

ezzanzzan committed:
test: 테스트 코드 DisplayName 수정

ezzanzzan committed:
test: LineIntegrationTest 수정

- LineStep 클래스 추가
- LineIntegrationTest 중복 코드 제거

ezzanzzan committed:
test: StationIntegrationTest 수정

- StationStep 클래스 추가
- StationIntegrationTest 중복 코드 제거

ezzanzzan committed:
test: SectionIntegrationTest 수정

- SectionStep 클래스 추가
- SectionIntegrationTest 중복 코드 제거

ezzanzzan committed:
test: StationStep 불필요한 코드 제거

Gomding commented on src/test/java/subway/integration/SectionIntegrationTest.java:
```diff
@@ -0,0 +1,134 @@
+
+@DisplayName("지하철 구간 관련 기능")
+@SuppressWarnings("NonAsciiCharacters")
+public class SectionIntegrationTest extends IntegrationTest {
```
API 통합 테스트에서는 일부분의 예외 상황에 대한 테스트만 있어보여요

단위테스트에서만 검증해도 괜찮을까요? 🤔

Gomding reviewed (approved):
안녕하세요 토리!
1단계 깔끔하게 구현해주셨네요 👍 
이번 단계는 이만 Merge 해도 되겠습니다.
다음 단계도 화이팅입니다 🔥 

추가로 궁금한 점 있으시면 언제든 편하게 DM 남겨주세요
